ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java

```diff
@@ -995,4 +1003,208 @@ void markStartTime() {
       callbackTracker.markOperationStart();
     }
   }
+
+  /**
+   * Injects target {@link Account} and {@link Container} for PUT requests. This method also ensures required headers
+   * are present for old and new PUT requests. A new PUT request is the one that carry both the {@code x-ambry-target-account}
+   * and {@code x-ambry-target-container} headers, otherwise it is not a new PUT request.
+   * @param restRequest The Put {@link RestRequest}.
+   * @throws RestServiceException
+   */
+  private void injectAccountAndContainerForPutRequest(RestRequest restRequest) throws RestServiceException {
+    accountAndContainerSanityCheck(restRequest);
+    if (getTargetAccountNameFromHeader(restRequest) != null || getTargetContainerNameFromHeader(restRequest) != null) {
+      ensureRequiredHeadersOrThrow(restRequest, requiredAmbryHeadersForNewPutRequest);
+      frontendMetrics.newPutRequestRate.mark();
+      injectAccountAndContainerForNewPutRequest(restRequest);
+    } else {
+      ensureRequiredHeadersOrThrow(restRequest, requiredAmbryHeadersForOldPuRequest);
+      frontendMetrics.oldPutRequestRate.mark();
+      injectAccountAndContainerForOldPutRequest(restRequest);
+    }
+  }
+
+  /**
+   * Injects {@link Account} and {@link Container} for the old put request, which does not carry account or container
+   * header.
+   * @param restRequest The {@link RestRequest} to inject {@link Account} and {@link Container} object.
+   * @throws RestServiceException if either of {@link Account} or {@link Container} object could not be identified.
+   */
+  private void injectAccountAndContainerForOldPutRequest(RestRequest restRequest) throws RestServiceException {
+    String accountName = getServiceId(restRequest);
+    Account targetAccount = accountService.getAccountByName(accountName);
+    if (targetAccount == null) {
+      frontendMetrics.oldPutUnrecognizedAccountCount.inc();
+      // @todo The check and config can be removed once every existing ambry user has been migrated to a valid account,
+      // @todo and we want to reject the PUT requests with unidentifiable account name.
+      if (frontendConfig.frontendShouldProhibitOldPutWithUnrecognizedAccountAndContainer) {
+        throw new RestServiceException("Account cannot be found for accountName=" + accountName + " in old put request",
+            RestServiceErrorCode.InvalidAccount);
+      } else {
+        logger.warn(
+            "Account cannot be found for accountName={} in old put request. Setting targetAccount to UNKNOWN_ACCOUNT",
+            accountName);
+        targetAccount = Account.UNKNOWN_ACCOUNT;
+      }
+    }
+    Container targetContainer;
+    boolean isBlobPrivate = isPrivate(restRequest.getArgs());
```

Comment:
It seems a bit unstraigforward and I could not find an easy way to reuse the value, unless we change the `buildBlobProperties` in `RestUtils`.